### ISSUE TITLE
[Epic] Removing clang from the Cilium agent image ('clang-free')

### ISSUE THREAD

**Issue by ti-mo** [cilium/cilium#27446]
This issue is for tracking work on removing the LLVM compilation toolchain from the Cilium agent runtime image and shipping prebuilt eBPF ELFs embedded in the agent binary. In short, this will provide a more natural developer experience (`if (feature_foo) { ... }` instead of `#ifdef FEATURE_FOO`), avoids high CPU consumption due to datapath (re)compilations during upgrades and reconfigs, allows for using real datapath ELFs in tests, and is a massive step in general towards shipping a distroless container image, the benefits of which won't be outlined here.

For anyone wishing to follow along, see the [clang-free milestone](https://github.com/cilium/cilium/milestone/45).

**Comment by aanm** [comment]
Will this issue solve
https://github.com/cilium/cilium/issues/21907 and https://github.com/cilium/cilium/pull/27829?

**Comment by ti-mo** [comment]
@aanm No. :sweat_smile: We'll always need clang, the goal is to build the datapath when building Cilium and not having to recompile it after.

**Comment by joestringer** [comment]
Notes from the Cilium community meeting discussion today on the latest of this effort (cc @jrife)

- Current work is swapping out compile time variables with runtime variables in BPF
- Recently merged work was some code to auto generate Go structs for datapath fields ("dpgen" / skeleton / scaffolding)
- Next step refactoring to make all code paths compile into the datapath with default values to determine what to call / not call. Replace "#ifdef ___" with "if (...) {...}"
- Easiest step to start is to just pick one variable and see if it can be converted into a runtime variable
